File: beast-arena-app/src/engine/SupplyDrop.ts
import { SupplyDrop, SupplyDropItem, FighterState, GAME_CONFIG, Buff } from './types';

const SUPPLY_ITEMS: SupplyDropItem[] = [
  { id: 'iron_gloves', type: 'weapon', name: 'Găng Tay Sắt', value: 0.3, duration: 600 },
  { id: 'nunchaku', type: 'weapon', name: 'Côn Nhị Khúc', value: 0.5, duration: 300 },
  { id: 'grilled_meat', type: 'heal', name: 'Thịt Nướng', value: 200, duration: 0 },
  { id: 'energy_drink', type: 'speed', name: 'Thuốc Tăng Lực', value: 3, duration: 480 },
  { id: 'shoulder_armor', type: 'weapon', name: 'Giáp Vai', value: 0.2, duration: 900 },
];

const SPAWN_INTERVAL_MIN = 20 * 60; // 20s in frames
const SPAWN_INTERVAL_MAX = 30 * 60; // 30s in frames
const WARNING_FRAMES = 9 * 60;      // 9 seconds warning

let dropIdCounter = 0;

export function createSupplyDropManager() {
  let nextSpawnFrame = randomBetween(SPAWN_INTERVAL_MIN, SPAWN_INTERVAL_MAX);
  const drops: SupplyDrop[] = [];

  function update(currentFrame: number): SupplyDrop | null {
    let newDrop: SupplyDrop | null = null;

    // Check if it's time to spawn a new drop
    if (currentFrame >= nextSpawnFrame - WARNING_FRAMES && !drops.some((d) => !d.active && !d.claimed)) {
      // Schedule warning
      const item = SUPPLY_ITEMS[Math.floor(Math.random() * SUPPLY_ITEMS.length)];
      const drop: SupplyDrop = {
        id: `drop_${++dropIdCounter}`,
        item: { ...item },
        position: {
          x: randomBetween(150, GAME_CONFIG.stageWidth - 150),
          y: GAME_CONFIG.groundY,
        },
        spawnFrame: nextSpawnFrame,
        warningFrame: nextSpawnFrame - WARNING_FRAMES,
        active: false,
        claimed: false,
      };
      drops.push(drop);
      newDrop = drop;
    }

    // Activate drops that have reached spawn frame
    for (const drop of drops) {
      if (!drop.active && !drop.claimed && currentFrame >= drop.spawnFrame) {
        drop.active = true;
      }
    }

    // Schedule next spawn
    if (currentFrame >= nextSpawnFrame) {
      nextSpawnFrame = currentFrame + randomBetween(SPAWN_INTERVAL_MIN, SPAWN_INTERVAL_MAX);
    }

    return newDrop;
  }

  function checkPickup(fighter: FighterState): SupplyDropItem | null {
    for (const drop of drops) {
      if (!drop.active || drop.claimed) continue;
      const dist = Math.abs(fighter.position.x - drop.position.x);
      if (dist < 50 && Math.abs(fighter.position.y - drop.position.y) < 60) {
        drop.claimed = true;
        drop.active = false;
        return drop.item;
      }
    }
    return null;
  }

  function getActiveDrops(): SupplyDrop[] {
    return drops.filter((d) => !d.claimed);
  }

  function reset(): void {
    drops.length = 0;
    nextSpawnFrame = randomBetween(SPAWN_INTERVAL_MIN, SPAWN_INTERVAL_MAX);
  }

  return { update, checkPickup, getActiveDrops, reset };
}

/** Apply a supply drop item to a fighter */
export function applySupplyItem(fighter: FighterState, item: SupplyDropItem): void {
  switch (item.type) {
    case 'heal':
      fighter.hp = Math.min(fighter.maxHp, fighter.hp + item.value);
      break;
    case 'weapon':
      fighter.activeBuffs.push({
        id: item.id,
        type: 'damage',
        value: item.value,
        remainingFrames: item.duration,
      });
      break;
    case 'speed':
      fighter.activeBuffs.push({
        id: item.id,
        type: 'speed',
        value: item.value,
        remainingFrames: item.duration,
      });
      break;
  }
}

function randomBetween(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

File: beast-arena-server/internal/game/supply_drop.go
package game

import (
	"math/rand"

	"github.com/google/uuid"
)

// SupplyDropType defines what a supply drop contains
type SupplyDropItem struct {
	ID    string  `json:"id"`
	Name  string  `json:"name"`
	Type  string  `json:"type"` // heal, damage_boost, speed_boost, shield, ult_charge
	Value float64 `json:"value"`
	Duration int  `json:"duration"` // frames (0 = instant)
}

var SupplyDropPool = []SupplyDropItem{
	{ID: "heal_small", Name: "Small Heal", Type: "heal", Value: 50, Duration: 0},
	{ID: "heal_large", Name: "Large Heal", Type: "heal", Value: 100, Duration: 0},
	{ID: "damage_boost", Name: "Damage Boost", Type: "damage_boost", Value: 1.3, Duration: 600}, // 10s at 60fps
	{ID: "speed_boost", Name: "Speed Boost", Type: "speed_boost", Value: 1.5, Duration: 480},     // 8s
	{ID: "shield", Name: "Shield", Type: "shield", Value: 50, Duration: 0},                       // absorbs 50 damage
	{ID: "ult_charge", Name: "Ultimate Charge", Type: "ult_charge", Value: 30, Duration: 0},       // +30% ult gauge
}

// SpawnSchedule determines when supply drops appear
type SpawnSchedule struct {
	minInterval int // minimum frames between spawns
	maxInterval int
	nextSpawn   int // frame number for next spawn
}

// NewSpawnSchedule creates a scheduler
func NewSpawnSchedule() *SpawnSchedule {
	return &SpawnSchedule{
		minInterval: 900,  // 15 seconds at 60fps
		maxInterval: 1800, // 30 seconds
		nextSpawn:   600,  // First drop at 10 seconds
	}
}

// CheckSpawn returns a new supply drop if it's time
func (s *SpawnSchedule) CheckSpawn(currentFrame int) *SupplyDrop {
	if currentFrame < s.nextSpawn {
		return nil
	}

	// Schedule next spawn
	s.nextSpawn = currentFrame + s.minInterval + rand.Intn(s.maxInterval-s.minInterval)

	// Random drop item
	item := SupplyDropPool[rand.Intn(len(SupplyDropPool))]

	// Random position (center area of stage)
	x := 300 + rand.Float64()*680 // between 300-980

	return &SupplyDrop{
		ID:       uuid.New().String(),
		Type:     item.Type,
		ItemID:   item.ID,
		Position: Vector2D{X: x, Y: 0},
		SpawnAt:  currentFrame,
		Warning:  currentFrame - 540, // 9 second warning
		Claimed:  false,
	}
}

// ClaimDrop applies the supply drop effect to a fighter
func ClaimDrop(drop *SupplyDrop, fighter *FighterState) {
	if drop.Claimed {
		return
	}
	drop.Claimed = true

	item := findItem(drop.ItemID)
	if item == nil {
		return
	}

	switch item.Type {
	case "heal":
		fighter.HP += int(item.Value)
		if fighter.HP > fighter.MaxHP {
			fighter.HP = fighter.MaxHP
		}
	case "ult_charge":
		fighter.UltGauge += item.Value
		if fighter.UltGauge > 100 {
			fighter.UltGauge = 100
		}
	case "damage_boost", "speed_boost":
		fighter.ActiveBuffs = append(fighter.ActiveBuffs, Buff{
			ID:       item.ID,
			Type:     item.Type,
			Value:    item.Value,
			Duration: item.Duration,
		})
	case "shield":
		fighter.ActiveBuffs = append(fighter.ActiveBuffs, Buff{
			ID:       "shield",
			Type:     "shield",
			Value:    item.Value,
			Duration: 99999, // until depleted
		})
	}
}

func findItem(id string) *SupplyDropItem {
	for _, item := range SupplyDropPool {
		if item.ID == id {
			return &item
		}
	}
	return nil
}
